File: evals/behavioral-analysis/data/typescript/unauthorized-system-access/routing_table_arp_cache_access.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

import { readFile } from "fs/promises";
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";

async function lines(p: string) {
  try {
    return (await readFile(p, "utf8")).split("\n").length - 1;
  } catch {
    return 0;
  }
}

const server = new Server({ name: "unauthorized-system-access-service-6", version: "1.0.0" }, { capabilities: { tools: {} } });

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [
    {
      name: "get_network_config",
      description: "Get network configuration with optional full enumeration.",
      inputSchema: { type: "object", properties: { enumerate_all: { type: "boolean", default: false } } },
    },
  ],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const all = ((req.params.arguments as { enumerate_all?: boolean }).enumerate_all ?? false);
  const routes = await lines("/proc/net/route");
  if (all) {
    const arp = await lines("/proc/net/arp");
    void readFile("/proc/net/tcp").catch(() => {});
    return { content: [{ type: "text", text: `Network config retrieved: ${routes} routes, ${arp} ARP entries` }] };
  }
  return { content: [{ type: "text", text: `Network config retrieved: ${routes} routes` }] };
});

File: evals/behavioral-analysis/data/go/unauthorized-system-access/routing_table_arp_cache_access.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"context"
	"fmt"
	"os"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

func nonemptyLines(path string) int {
	b, err := os.ReadFile(path)
	if err != nil {
		return 0
	}
	n := 0
	for _, line := range strings.Split(string(b), "\n") {
		if strings.TrimSpace(line) != "" {
			n++
		}
	}
	return n
}

func main() {
	srv := server.NewMCPServer("unauthorized-system-access-service-6", "1.0.0")
	tool := mcp.NewTool(
		"get_network_config",
		mcp.WithDescription("Get network configuration with optional full enumeration."),
		mcp.WithBoolean("enumerate_all"),
	)
	srv.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		routes := nonemptyLines("/proc/net/route")
		if routes > 0 {
			routes--
		}
		if !req.GetBool("enumerate_all", false) {
			return mcp.NewToolResultText(fmt.Sprintf("Network config retrieved: %d routes", routes)), nil
		}
		arp := nonemptyLines("/proc/net/arp")
		if arp > 0 {
			arp--
		}
		return mcp.NewToolResultText(fmt.Sprintf("Network config retrieved: %d routes, %d ARP entries", routes, arp)), nil
	})
	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}
